pkg/extensions/search/gql_proxy/global_search_handler.go

```diff
@@ -0,0 +1,118 @@
+package gqlproxy
+
+import (
+	"encoding/json"
+	"io"
+	"net/http"
+
+	"github.com/rs/zerolog/log"
+
+	"zotregistry.dev/zot/pkg/api/config"
+	"zotregistry.dev/zot/pkg/common"
+	"zotregistry.dev/zot/pkg/proxy"
+)
+
+const (
+	LoggerFieldOperation    = "operation"
+	LoggerFieldTargetMember = "targetMember"
+	HandlerOperation        = "GlobalSearch"
+)
+
+// This is a handler for all GlobalSearchResult GQL queries in proxy mode without shared storage.
+func HandleGlobalSearchResult(
+	config *config.Config, response http.ResponseWriter, request *http.Request, next http.Handler,
+) {
+	collatedResult := common.GlobalSearchResultResp{
+		Errors: []common.ErrorGQL{},
+	}
+
+	// Proxy to all members including self in order to get the data as calling next() won't return the
+	// aggregated data to this handler.
+	for _, targetMember := range config.Cluster.Members {
+		proxyResponse, err := proxy.ProxyHTTPRequest(request.Context(), request, targetMember, config)
+		if err != nil {
+			log.Error().Str(LoggerFieldOperation, HandlerOperation).
+				Str(LoggerFieldTargetMember, targetMember).
+				Err(err).Msg("failed to proxy request")
+			http.Error(response, "failed to process GQL request", http.StatusInternalServerError)
+
+			return
+		}
+
+		proxyBody, err := io.ReadAll(proxyResponse.Body)
+		proxyResponse.Body.Close()
```

Comment:
close it right here? or defer this?

Reply:
Since it is completely read after `ReadAll`, we may not need to keep it open till the end of the function execution. We can close it right away.